### PR TITLE
fabtests: Add benchmark option for `FI_OPT_MAX_MSG_SIZE`

### PR DESCRIPTION
**Pull request by darrylabbate**
This adds an option (`-X`) for benchmark tests; corresponding to the endpoint options `FI_OPT_MAX_MSG_SIZE`. `fi_setopt()` is called during EP enable with the value requested by the user.

This also includes a few related minor changes to the EFA provider.

### CODE REVIEW

File: fabtests/benchmarks/benchmark_shared.c [hunk 1]
```diff
@@ -79,6 +82,7 @@ void ft_benchmark_usage(void)
 	FT_PRINT_OPTS_USAGE("-v", "enables data_integrity checks");
 	FT_PRINT_OPTS_USAGE("-k", "force prefix mode");
 	FT_PRINT_OPTS_USAGE("-j", "maximum inject message size");
+	FT_PRINT_OPTS_USAGE("-X", "maximum untagged message size");
```
Comment (shijin-aws) [inline fabtests/benchmarks/benchmark_shared.c:85]: I think using a general `-X` only for untagged message size is not desired. We can introduce a long option `--max-msg-size` for it.

File: fabtests/common/shared.c [hunk 2]
```diff
@@ -1427,6 +1427,14 @@ int ft_enable_ep(struct fid_ep *bind_ep, struct fid_eq *bind_eq, struct fid_av *
 		FT_EP_BIND(bind_ep, bind_rma_cntr, flags);
 	}
 
+	if (opts.max_msg_size) {
+		ret = fi_setopt(&bind_ep->fid, FI_OPT_ENDPOINT, FI_OPT_MAX_MSG_SIZE, &opts.max_msg_size, sizeof opts.max_msg_size);
+		if (ret) {
```
Comment (shijin-aws) [inline fabtests/common/shared.c:1432]: What will happen if provider doesn't support such option? We should ignore such ENOSUPPORT error

**Review by shijin-aws — CHANGES_REQUESTED** [review]

**Comment by shijin-aws** [comment]
Do you need another commit to add such test to pytest?

**Comment by darrylabbate** [comment]
> Do you need another commit to add such test to pytest?

Are you asking for a default pytest? I figure we can have a follow-up PR for some kind of EFA zero-copy fabtest.

**Comment by shijin-aws** [comment]
> Are you asking for a default pytest? I figure we can have a follow-up PR for some kind of EFA zero-copy fabtest.

I prefer to just have it in this PR so we can make it tested by AWS CI. It should be a straightforward incremental change, like what we have in `test_rdm.py`

File: fabtests/common/shared.c [hunk 3]
```diff
@@ -1427,6 +1427,14 @@ int ft_enable_ep(struct fid_ep *bind_ep, struct fid_eq *bind_eq, struct fid_av *
 		FT_EP_BIND(bind_ep, bind_rma_cntr, flags);
 	}
 
+	if (opts.max_msg_size) {
+		ret = fi_setopt(&bind_ep->fid, FI_OPT_ENDPOINT, FI_OPT_MAX_MSG_SIZE, &opts.max_msg_size, sizeof opts.max_msg_size);
+		if (ret && ret != -FI_ENOPROTOOPT) {
```
Comment (shijin-aws) [inline fabtests/common/shared.c:1432]: I think it's a bug in the man page. In this case we should return `-FI_EOPNOTSUPP`

Comment (shijin-aws) [reply fabtests/common/shared.c:1432]: Nvm, I think FI_ENOPROTOOPT is OK and that's what efa and shm returns today

File: fabtests/pytest/efa/test_rdm.py [hunk 4]
```diff
@@ -109,3 +109,8 @@ def test_rdm_pingpong_1G(cmdline_args, completion_semantic):
     efa_run_client_server_test(cmdline_args, "fi_rdm_pingpong -W 1", 2,
                                completion_semantic=completion_semantic, message_size=1073741824,
                                memory_type="host_to_host", warmup_iteration_type=0)
+
+@pytest.mark.functional
```
Comment (shijin-aws) [inline fabtests/pytest/efa/test_rdm.py:113]: can you add rdm_bw as well?

**Review by shijin-aws — APPROVED** [review]

**Comment by shijin-aws** [comment]
@j-xiong can you review https://github.com/ofiwg/libfabric/pull/10147/commits/fa178bb1fe878de3be948304e3504a9b8da92ef8

**Review by j-xiong — APPROVED** [review]

File: fabtests/pytest/efa/test_rdm.py [hunk 5]
```diff
@@ -109,3 +109,13 @@ def test_rdm_pingpong_1G(cmdline_args, completion_semantic):
...
+def test_rdm_pingpong_zcpy_recv(cmdline_args, memory_type, message_size):
+    efa_run_client_server_test(cmdline_args, "fi_rdm_pingpong --max-msg-size 8192",
+                               "short", "transmit_complete", memory_type, message_size)
+
+@pytest.mark.functional
+def test_rdm_bw_zcpy_recv(cmdline_args, memory_type, message_size):
```
Comment (shijin-aws) [inline fabtests/pytest/efa/test_rdm.py:119]: The `message_size` fixture will have msg size larger than 8192, you need to customize it to make it smaller than 8192

**Comment by shijin-aws** [comment]
I opened a new PR that reworked the last commit of this PR https://github.com/ofiwg/libfabric/pull/10168

**Pull request closed without merging**